File: 22FALL/110.balanced-binary-tree.py
#
# @lc app=leetcode id=110 lang=python3
#
# [110] Balanced Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        def balance(node):
            if not node:
                return 0
            lh = balance(node.left)
            rh = balance(node.right)
            
            if abs(lh-rh) > 1 or lh == -1 or rh == -1:
                return -1
            return max(lh, rh) + 1
        return True if balance(root) > -1 else False

# @lc code=end

File: VSCode_CPP/110.balanced-binary-tree.cpp
/*
 * @lc app=leetcode id=110 lang=cpp
 *
 * [110] Balanced Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int checkHeight(TreeNode* node, bool& ans) {
        if (ans == false) return -1;
        if (!node) return 0;
        int lheight = checkHeight(node->left, ans);
        int rheight = checkHeight(node->right, ans);
        if (abs(lheight-rheight) > 1) ans = false;
        return max(lheight, rheight)+1;
    }
    bool isBalanced(TreeNode* root) {
        bool ans = true;
        int a = checkHeight(root, ans);
        return ans;
    }
};
// @lc code=end

